spec/aws-record/record/dirty_tracking_spec.rb

```diff
@@ -0,0 +1,240 @@
+
+      instance.mykey = SecureRandom.uuid
+      expect(instance.mykey_dirty?).to be true
+
+      instance.mykey = nil
+      expect(instance.mykey_dirty?).to be false
+    end
+  end
+
+  describe '#[attribute]_dirty!' do 
+
+    before(:each) do 
+      instance.mykey = "Alex"
+      instance.clean!
+    end
+
+    it "should mark the attribute as dirty" do 
+      instance.mykey << 'i'
+      expect(instance.dirty?).to be false
+
+      instance.mykey_dirty!
+      expect(instance.dirty?).to be true 
+      expect(instance.mykey_was).to eq 'Alexi'
+      expect(instance.mykey).to eq 'Alexi'
+
+      instance.mykey << 's'
+      expect(instance.mykey_was).to eq 'Alexi'
+      expect(instance.mykey).to eq 'Alexis'
+    end
+
+  end
+
+  describe '#[attribute]_was' do 
+
+    it "should return the last known clean value" do 
+      expect(instance.mykey_was).to be nil
+
+      instance.mykey = SecureRandom.uuid
+      expect(instance.mykey_was).to be nil
+    end
+
+  end
+
+  describe "#clean!" do 
+
+    it "should mark the record as clean" do 
+      instance.mykey = SecureRandom.uuid
+      expect(instance.dirty?).to be true
+      expect(instance.mykey_was).to be nil
+
+      instance.clean!
+      expect(instance.dirty?).to be false
+      expect(instance.mykey_was).to eq instance.mykey
+    end
+
+  end
+
+  describe '#dirty' do 
+
+    it "should return an array of dirty attributes" do 
+      expect(instance.dirty).to match_array []
+
+      instance.mykey = SecureRandom.uuid
+      expect(instance.dirty).to match_array [:mykey]
+
+      instance.body = SecureRandom.uuid
+      expect(instance.dirty).to match_array [:mykey, :body]
+    end 
+
+  end
+
+  describe '#dirty?' do 
+
+    it "should return whether the record is dirty or clean" do 
+      expect(instance.dirty?).to be false
+
+      instance.mykey = SecureRandom.uuid
+      expect(instance.dirty?).to be true
+    end
+
+  end
+
+  describe "#reload!" do 
+
+    before(:each) do 
+      klass.configure_client(client: stub_client)
+    end
+
+    let(:reloaded_instance) {
+      item = klass.new
+      item.mykey = SecureRandom.uuid
+      item.body = SecureRandom.uuid
+      item.clean!
+      item 
+    }
+
+    it 'can reload an item using find' do 
+      expect(klass).to receive(:find).with({ mykey: reloaded_instance.mykey }).and_return(reloaded_instance)
+
+      instance.mykey = reloaded_instance.mykey
+      instance.body = SecureRandom.uuid
+
+      instance.reload!
+
+      expect(instance.body).to eq reloaded_instance.body
+    end
+
+    it 'raises an error when find returns nil' do 
+      instance.mykey = SecureRandom.uuid
+
+      expect(klass).to receive(:find).with({ mykey: instance.mykey }).and_return(nil)
+
+      expect { instance.reload! }.to raise_error Aws::Record::Errors::NotFound
+    end
+
+    it "should mark the item as clean" do 
+      instance.mykey = SecureRandom.uuid
+      expect(instance).to receive(:clean!)
```

Comment:
I don't think we should test this in this way (because it's implementation dependent on calling `#clean!`. Rather, could we validate that a reloaded object doesn't register as dirty?